src/pmu.rs

```diff
@@ -0,0 +1,303 @@
+            let out = format!(
+                "\n{}\nDelete?",
+                serde_json::to_string_pretty(&perf_list[index])?
+            );
+            if Confirm::new(&out).with_default(true).prompt()? {
+                perf_list.remove(index);
+            }
+        }
+        if !Confirm::new("Remove more?").with_default(false).prompt()? {
+            break;
+        }
+    }
+    Ok(perf_list.to_vec())
+}
+
+pub fn create_pmu_config(cpmu: &CustomPMU) -> Result<()> {
+    let mut pmu_file = PathBuf::from("aperf_custom_pmu.json");
+    if let Some(f) = &cpmu.pmu_file {
+        pmu_file = PathBuf::from(f);
+    }
+    #[cfg(target_arch = "aarch64")]
+    let events: Vec<NamedCtr> = serde_json::from_slice(arm64_perf_list::GRV_EVENTS)?;
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    let events: Vec<NamedCtr> = serde_json::from_slice(x86_perf_list::INTEL_EVENTS)?;
+    #[cfg(target_arch = "aarch64")]
+    let platform = "Graviton";
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    let platform = "Intel";
+    println!(
+        "\nAperf PMU Event structure (Ex: {})\n{}",
+        platform,
+        serde_json::to_string_pretty(&events[0])?
+    );
+    println!(
+        "\nPlease enter your custom PMU details. Only hex values (0x) for 'config' will work.\n"
+    );
+    let events = add_events(Vec::new())?;
+    let f = std::fs::File::create(&pmu_file)?;
+    serde_json::to_writer_pretty(f, &events)?;
+    println!(
+        "\nCustom PMU config generated at: {:?}. Use this with 'aperf record --pmu-file {:?}'.",
+        pmu_file, pmu_file
+    );
+    Ok(())
+}
+
+pub fn get_config(choice: &str, cpmu: &CustomPMU) -> Result<Vec<NamedCtr>> {
+    if choice == "User provided" {
+        if let Some(f) = &cpmu.pmu_file {
+            let file = std::fs::File::open(PathBuf::from(f))?;
+            return Ok(serde_json::from_reader(&file)?);
+        } else {
+            println!("No custom config file provided.");
+            return Err(PDError::PMUCustomFileNotFound.into());
+        }
+    };
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    match choice {
+        "Intel" => to_events(x86_perf_list::INTEL_EVENTS),
+        "Intel Sapphire Rapids" => {
+            form_events_map(x86_perf_list::INTEL_EVENTS, x86_perf_list::SPR_CTRS)
+        }
+        "Intel Icelake" => form_events_map(x86_perf_list::INTEL_EVENTS, x86_perf_list::ICX_CTRS),
+        "AMD" => to_events(x86_perf_list::AMD_EVENTS),
+        "AMD Genoa" => form_events_map(x86_perf_list::AMD_EVENTS, x86_perf_list::GENOA_CTRS),
+        "AMD Milan" => form_events_map(x86_perf_list::AMD_EVENTS, x86_perf_list::MILAN_CTRS),
+        _ => Ok(Vec::new()),
+    }
+    #[cfg(target_arch = "aarch64")]
+    match choice {
+        "Graviton" => to_events(arm64_perf_list::GRV_EVENTS),
+        _ => Ok(Vec::new()),
+    }
+}
+
+pub fn modify_existing_config(cpmu: &CustomPMU) -> Result<()> {
```

Comment:
Personally, I don't think anyone will be using this input method.  Just ask them to use a text editor and provide documentation on the format.  I'd suggest deleting the modification routines, but I would suggest keeping the ability to dump what _would_ be collected to a file for the user to modify.

Reply:
The issue I see is that a user will create a file while eyeballing the format. Any errors will only be identified when passing the file into aperf which errors out. I'd rather prevent the user from creating an invalid file in the first place.

Reply:
you can't stop me from making an invalid file😈! It would be more helpful to provide me a way to validate/parse the file.  And of course if they are editing the file, they are probably going to be able to re-edit it to fix any errors that pop up.